multicluster/controllers/multicluster/member/serviceexport_controller.go

```diff
@@ -231,19 +238,22 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 			Namespace: req.Namespace,
 		},
 	}
-	readyAddresses := 0
-	err = r.Client.Get(ctx, req.NamespacedName, eps)
-	if err == nil {
-		for _, subsets := range eps.Subsets {
-			readyAddresses = readyAddresses + len(subsets.Addresses)
+
+	var hasReadyEndpoints bool
+	if r.endpointSliceEnabled {
+		eps.Subsets, hasReadyEndpoints, err = r.getSubsetsFromEndpointSlice(ctx, req)
+		if err != nil {
+			return ctrl.Result{}, err
 		}
 	} else {
-		if !apierrors.IsNotFound(err) {
+		hasReadyEndpoints, err = r.checkSubsetsFromEndpoint(ctx, req, eps)
+		if err != nil {
 			klog.ErrorS(err, "Failed to get Endpoints", req.String())
 			return ctrl.Result{}, err
 		}
 	}
-	if readyAddresses == 0 || apierrors.IsNotFound(err) {
+
+	if !hasReadyEndpoints {
 		if err := cleanup(); err != nil {
```

Comment:
I just checked the cleanup() code, and got a question - if svcInstalled/epInstalled is false, do we still 
 need to call `handleService/EndpointsDeleteEvent`? Why?

Reply:
Yes, the main purpose of svcInstalled and epsInstalled is to avoid unnecessary API call to leader. but this might be always false when controller is just launched or restarted but no cache.  We need to continue to handle it. The worst case is to call leader's API but actually it's no-ops for an update action, for deletion action, we will get not found error but ignore it. For creation.

Reply:
For the restart case, stale controller cannot handle the cleanup?

Reply:
stale controller check local ServiceExport list, and clean up remote ResoureExports if there is no corresponding ServiceExport(deleted) any more. Here is the case for an existing ServiceExport but has no Endpoints.
During a restart case, svcInstalled is false, but there might be no available addresses anymore during mc-controller restart, so we still need to handle this kind of case which is not covered by stale controller.

Reply:
Could we add a comment about this case?

Reply:
Done